### ISSUE TITLE
Parallelization of computations and the SymEngine library

### ISSUE THREAD

**Issue by UInSomnia** [symengine/symengine#2079]
Hello!

I am writing a neural network constructor on
JAVA C++,
Linux OS,
CMake collector,
IDE QtCreator (I don't use the Qt library itself, it's just convenient to write code in QtCreator)
SymEngine is a library that allows you to perform symbolic calculations. For example, store 2 * x + 4 * y. And then take the derivative and substitute the values instead of the unknowns. It is very convenient in neural network training (taking, storing derivatives and setting values in them)

The code I provide below works fine in single-threaded mode.
I decided to parallelize finding the gradient over the unknowns into n-threads.
For example, we have unknowns x1, x2, x3, x4. If we divide into two streams, the first one counts the derivatives of x1 and x2, the second one simultaneously counts the derivatives of x3 and x4. Then we measure their result.

I've been struggling for hours on how to set this up. SymEngine has a basic class called Basic. I called the basic wrapped in a smart pointer (RCP) SE_Basic. Different mathematical entities can act as Basic - int, double, sum, product, sine, cosine, and so on. SymEngine stores the expression as a graph. Where we have access to its root node, which is stored in SE_Basic - this is a reflection of our expression

Here is the code. An attempt to parallelize the search for gradients by pre-calculated derivatives (dir)

```
void compute_gradients(
	const size_t index_thread,
	const size_t index_start,
	const size_t index_finish,
	const std::vector<SE_Basic> &xs_names,
	const map_basic_basic &fun_loss_derivatives,
	const map_basic_basic &memory_values,
	std::vector<SE_Basic> &grad,
	std::mutex &mtx)
{
	std::cout << "Start compute_gradients() in "
			  << index_thread << " thread\n";
	std::cout.flush();
	
	const size_t count = index_finish - index_start + 1;
	std::vector<SE_Basic> grad_temp(count);
	
	for (size_t i = index_start; i <= index_finish; ++i)
	{
		
		const SE_Basic &name = xs_names[i];
		
		const SE_Basic &der = fun_loss_derivatives.at(name);
		
		const std::string str = der->__str__();
		const size_t len = str.length();
		
		std::cout << index_thread << " der length: " << len << "\n";
		std::cout.flush();
		
		// const SE_Basic &der_value = der->subs(memory_values);
		
		// std::cout << index_thread << " der_value OK\n";
		// std::cout << der_value << "\n";
		// std::cout.flush();
		
		// grad_temp[i - index_start] = std::move(der_value);
	}
	
	// std::lock_guard<std::mutex> lock(mtx);
	// for (size_t j = index_start; j <= index_finish; ++j)
	// {
	//     grad[j] = grad_temp[j - index_start];
	// }
}

void train (/* args */)
{
	// ...
	
	// Threads
	std::mutex mtx;
	const size_t count_threads =
		std::thread::hardware_concurrency();
	std::cout << "count_threads: " << count_threads << "\n";
	std::cout.flush();
	threads = std::vector<std::unique_ptr<std::thread>>(count_threads);

	for (size_t t = 0u; t < count_threads; ++t)
	{
		const size_t start = t * (count_x / count_threads);
		const size_t end = (t == count_threads - 1) ?
			count_x - 1 : start + (count_x / count_threads);
		
		std::cout << start << " -> " << end << "\n";
		std::cout.flush();
		
		std::thread *ptr_thread = new std::thread(
			InSomnia::compute_gradients,
			t,
			start,
			end,
			std::cref(xs_names),
			std::cref(fun_loss_derivatives),
			std::cref(memory_values),
			std::ref(grad),
			std::ref(mtx));
						
		threads[t] = std::unique_ptr<std::thread>(ptr_thread);
		
	}

	std::cout << "Start join\n";
	std::cout.flush();

	for (std::unique_ptr<std::thread> &t : threads)
	{
		t->join();
	}

	std::cout << "Finish join\n";
	
	// ...
	
}
```

This is the basic representation of your world here. For example, "x1". This works fine
`const SE_Basic &name = xs_names[i];`
Here is the main idea from your life (Add). That's why it's a huge image of a derivative of one of the unknowns.
`const SE_Basic &der = fun_loss_derivatives.at (name);`
It's a tree. It is a constant amount of money.
This thing doesn't work with parallelization.

Any access to it arouses interest and outputs programs. I am analyzing the text
`const std::string str = der->__str__()`

Error description:
RTTI symbol not found for class 'SymEngine::Add'
The inferior stopped because it received a signal from the operating system.

Signal name : SIGABRT
Signal meaning : Aborted

I tried it in CMakeLists.txt add
`set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -frtti")`
It didn't help

Help me set up SymEngine for multithreaded execution
Thanks for the help!

**Comment by rikardn** [comment]
Was symengine built with thread safety?

To help better understand the issue it would be good if you could provide a minimal reproducible example that could be compiled and tested.

**Comment by UInSomnia** [comment]
@rikardn, Thank you for your reply!
Yes. A solution has been found
It was necessary to build SymEngine with the required CMake key
`cmake -DWITH_SYMENGINE_THREAD_SAFE=ON ..`

**Comment by rikardn** [comment]
Thanks for the feedback!
